Fix collapsing input queues and other window handling fixes and enhancements

tothxa opened the pull request:
**Type of change**
New feature

**Issue(s) closed**
Fixes #5605

Also fixes remembering the position of UniqueWindows (e.g. statistics and editor tools).

**To reproduce**
*If it's a bugfix:*
Steps to reproduce the behavior which cause the bug in master but not in your branch:

Production sites:

1. Open a productionsite window
2. Collapse the inputqueues
3. Click stop
4. See error

Other windows:

1. Open the general statistics window
2. Move it from its default position
3. Toggle it off from the Statistics menu or with its hotkey (default `A`)
4. Reopen it → it opens in the default position again
5. Move it again
6. Close it by right-clicking or with the close button
7. Reopen it →it opens in the position where it was closed – I believe this is the intended behaviour

**New behavior**
The collapsed state of inputqueue displays is stored in the building window's registry to remember it when the window is closed. (stopping/restarting practically closes and reopens the window) The state is also saved with the game, ~but currently it only works~ for open windows ~– that's why this is a draft.~ (like other window types)

As a side effect, construction sites now have the same collapsed state for both the building materials and the settings for the future production site. I think this actually makes sense.

Initial positioning of building (and ship and field action) windows is changed so that the window always opens near the clicked position, but if possible, with a clearance to allow seeing the surroundings.

**Possible regressions**
Handling, display and saving/loading of building and other windows.

tothxa committed:
Make building windows remember collapsed state of input queues

Noordfrees reviewed (commented):
Very elegant solution :) 
Needs a few documentary comments.

One thing I'm worried about is that it's all too easy to confuse `bool*` with `bool` and treat the pointer like the underlying boolean. Maybe replacing `bool` with an `enum class CollapsedState { kExpanded, kCollapsed }` would be safer?

Not tested yet.

tothxa committed:
Apply suggestions from code review

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

tothxa committed:
7 files were automatically formatted.

tothxa commented:
> Very elegant solution :)

Thank you. :) Actually I didn't have much choice though. :) The problem turned out to be much harder than I expected, and I started out with more complicated code in `BuildingWindow`, but as I went on to `InputQueueDisplay` it just fell into place in this form.

tothxa committed:
Merge branch 'master' into ware-prio-registry

tothxa committed:
address review:
 * use enum class instead of bool
 * add comments and variable names

tothxa committed:
fix updating collapsed state

tothxa committed:
'src/wui/inputqueuedisplay.cc' was automatically formatted.

tothxa commented:
> The state is also saved with the game, but currently it only works for open windows – that's why this is a draft.

I've checked the window state saving code and tested it, and it turns out that only the state of currently open windows is saved for other window types too. I think collapsed state is less important than say minimap state, so I shouldn't go out of my way to implement saving state for closed windows in this PR.

So this PR is ready after all.

tothxa committed:
clang-tidy: use same long names for init args in inputqueuedisplay.cc as in header

tothxa committed:
'src/wui/inputqueuedisplay.cc' was automatically formatted.

tothxa committed:
Merge branch 'master' into ware-prio-registry

Commit:
Fetched translations and updated catalogs.

frankystone commented:
Works very well :-)

Loading an older save game the behavior is always collapsed (if i remember correctly): E.g. when building a new tavern the state is collapsed or when enhancing a mine the state is collapsed although it was previously not collapsed.

Otherwise a very good enhancement, imho :+1:

tothxa committed:
make expanded state explicit on loading old saveames

tothxa committed:
Merge branch 'master' into ware-prio-registry

tothxa commented:
> Loading an older save game the behavior is always collapsed (if i remember correctly): E.g. when building a new tavern the state is collapsed or when enhancing a mine the state is collapsed although it was previously not collapsed.

I couldn't reproduce it, but added code to make sure windows are expanded when loaded from old games.

-----

(sorry for the botched master merge / rebase / push)

frankystone commented:
Found another thing: If a window with collapsed prio section is close to the right screen edge and you click stop/start the window jumps to left. I guess the window jumps the amount of pixels the ware priority sliders would have need.

tothxa committed:
Merge branch 'master' into ware-prio-registry

tothxa committed:
fix UniqueWindow position saving and restoring

tothxa commented:
> Found another thing: If a window with collapsed prio section is close to the right screen edge and you click stop/start the window jumps to left. I guess the window jumps the amount of pixels the ware priority sliders would have need.

Fixed, and on the side I also fixed saving and restoring positions of all `UniqueWindow`s like the various statistics windows and tool windows in the editor.

It stressed my abilities to hunt it down though. ;) I needed `gdb`'s help to find the culprits.

~Now I need to debug why building windows are still not reopened at the last position.~ *(scrapped, because it turned out to be annoying for most buildings)*

tothxa committed:
improve positioning of building-, ship-, and field action windows

tothxa commented:
Last commit does not work right in watch windows. I believe it's the same problem as #4689.

frankystone commented:
From my side commit [58aa436](https://github.com/widelands/widelands/pull/5785/commits/58aa43618bd9abf5abd36f14cb4658c629cc811a) works as expected. Maybe you should not try to solve too many bugs in this pr?

Anyway many thanks for your work!

tothxa commented:
> From my side commit [58aa436](https://github.com/widelands/widelands/pull/5785/commits/58aa43618bd9abf5abd36f14cb4658c629cc811a) works as expected.

Unfortunatley it was inconsistent, and for the wrong reason.

When new (including just reopened) building windows are first drawn, `move_out_of_the_way()` is called, which despite the name, moved them first to the centre of the main window, and then, if they covered the position of the mouse pointer (that was presumably clicked to bring up the window), were moved to just above/below the pointer, and then `move_inside_parent()` was called to make sure that the whole building window is inside the main window. 
But [58aa436](https://github.com/widelands/widelands/pull/5785/commits/58aa43618bd9abf5abd36f14cb4658c629cc811a) changed `move_inside_parent()` to restore the previous position, so instead of moving to centre then, if needed, out of the way but making sure it stays inside, it became either move to centre, or if centre covers pointer, to old position (regardless of old position covering the pointer).

So I tried the easy way, to always move to the old position if there was one, but as I wrote above, it was annoying when you opened and closed windows of minor buildings.

> Maybe you should not try to solve too many bugs in this pr?

I know, but stomping out one problem leads to the next one, so it's hard to stop. :)

Now it turns out that I must solve the initial sizing of collapsed building windows anyway, because it is needed for `move_out_of_the_way()` in either case to centre the window correctly, and because the mouse pointer is moved to the rightmost button, but then previously collapsed windows shrink under it, so it's left outside the window.

tothxa committed:
try to fix initial sizing of input queues

tothxa commented:
It seems to work now… Probably https://github.com/widelands/widelands/pull/5785/commits/66bb34fe6fd973ec71913ae9cdbdb1337eb0c987 both fixed and broke it. It should have kept `set_collapsed()` in the constructor too… I can't remember if I tested it this way back then…

tothxa committed:
Merge branch 'master' into ware-prio-registry

frankystone commented:
I think this is fine now.

tothxa committed:
Merge branch 'master' into ware-prio-registry

hessenfarmer reviewed (approved):
Code looks reasonable (although I don't understzand it fully).
Tested and working

tothxa committed:
Merge branch 'master' into ware-prio-registry

tothxa commented:
> Code looks reasonable (although I don't understzand it fully). Tested and working

Thank you for the review. Please feel free to point out where it may need explanatory comments.

hessenfarmer commented:
> > Code looks reasonable (although I don't understzand it fully). Tested and working
> 
> Thank you for the review. Please feel free to point out where it may need explanatory comments.    
   
the problem is not missing comments but lack of knowledge on my side ;-) I simply don't have all parts of our code understood completely.
